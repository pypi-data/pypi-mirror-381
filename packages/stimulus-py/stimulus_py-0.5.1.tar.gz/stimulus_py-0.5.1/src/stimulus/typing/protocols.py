"""Protocol definitions for stimulus models."""

from typing import Any, Protocol

import torch
from torch import Tensor


class StimulusModel(Protocol):
    """Protocol for stimulus models with batch and inference methods."""

    def batch(
        self,
        batch: dict[str, Tensor],
        optimizer: Any | None = None,
        **loss_dict: Any,
    ) -> tuple[Tensor, dict[str, Any]]:
        """Process a batch and return loss and metrics.

        Args:
            batch: Dictionary of input tensors
            optimizer: Optional optimizer for training
            **loss_dict: Additional loss function arguments

        Returns:
            Tuple of (loss tensor, metrics dictionary)
        """
        ...

    def train_batch(
        self,
        batch: dict[str, Tensor],
        optimizer: Any,
        **loss_dict: Any,
    ) -> tuple[Tensor, dict[str, Any]] | tuple[Tensor, dict[str, Any], dict[str, Any]]:
        """Process a training batch and return loss, metrics, and optionally per-sample data.

        Args:
            batch: Dictionary of input tensors
            optimizer: Optimizer for training
            **loss_dict: Additional loss function arguments

        Returns:
            Tuple of (loss tensor, metrics dictionary) or
            Tuple of (loss tensor, metrics dictionary, per-sample dictionary)
        """
        ...

    def inference(
        self,
        batch: dict[str, Tensor],
        **loss_dict: Any,
    ) -> tuple[Tensor, dict[str, Any]]:
        """Run inference on a batch and return loss and metrics.

        Args:
            batch: Dictionary of input tensors
            **loss_dict: Additional loss function arguments

        Returns:
            Tuple of (loss tensor, metrics dictionary)
        """
        ...

    def eval(self) -> None:
        """Set model to evaluation mode."""
        ...

    def to(self, device: torch.device) -> "StimulusModel":
        """Move model to device.

        Args:
            device: Target device

        Returns:
            Model on the target device
        """
        ...
