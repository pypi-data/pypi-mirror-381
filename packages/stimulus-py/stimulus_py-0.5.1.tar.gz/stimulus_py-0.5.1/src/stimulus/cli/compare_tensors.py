"""CLI module for comparing tensors."""

import logging
from collections import defaultdict

import polars as pl
import torch
from safetensors.torch import load_file

logger = logging.getLogger(__name__)


def cosine_similarity(tensor1: torch.Tensor, tensor2: torch.Tensor) -> torch.Tensor:
    """Compute the cosine similarity between two tensors.

    Args:
        tensor1: First tensor.
        tensor2: Second tensor.

    Returns:
        The cosine similarity between the two tensors.
    """
    flat_tensor1 = tensor1.flatten()
    flat_tensor2 = tensor2.flatten()
    return torch.nn.functional.cosine_similarity(flat_tensor1, flat_tensor2, dim=0)


def discrete_compare(tensor1: torch.Tensor, tensor2: torch.Tensor, threshold: float = 0.0) -> torch.Tensor:
    """Compute the discrete comparison between two tensors.

    Args:
        tensor1: First tensor.
        tensor2: Second tensor.

    Returns:
        The discrete comparison between the two tensors.
    """
    flat_tensor1 = tensor1.flatten()
    flat_tensor2 = tensor2.flatten()
    # for each element in the flat tensors, set to 1 if the element is greater than the threshold, otherwise set to 0
    flat_tensor1 = (flat_tensor1 > threshold).float()
    flat_tensor2 = (flat_tensor2 > threshold).float()
    # compute the dot product of the two tensors and divide by the number of elements in the tensors
    # Calculate element-wise equality (1 where both are same, 0 where different)
    # For binary tensors (0s and 1s), we want:
    # - 1 when both tensors have same value (both 0 or both 1)
    # - 0 when tensors have different values
    equality = 1.0 - torch.abs(flat_tensor1 - flat_tensor2)
    # Return the proportion of matching elements
    return torch.sum(equality) / flat_tensor1.numel()


def compare_tensors(
    tensor1: dict[str, torch.Tensor],
    tensor2: dict[str, torch.Tensor],
    mode: str = "cosine_similarity",
) -> dict[str, torch.Tensor]:
    """Compare tensors.

    Args:
        tensor1: First tensor.
        tensor2: Second tensor.
        mode: Mode to use for comparison.

    Returns:
        A dictionary of statistics comparing tensors
    """
    compare_modes: dict = {
        "cosine_similarity": cosine_similarity,
        "discrete_comparison": discrete_compare,
    }

    tensor_comparison: dict[str, torch.Tensor] = {}

    if mode not in compare_modes:
        raise ValueError(f"Invalid mode: {mode}, available modes: {compare_modes.keys()}")

    predictions_comparison: torch.Tensor = compare_modes[mode](tensor1["predictions"], tensor2["predictions"])
    tensor_comparison[f"{mode}"] = predictions_comparison
    common_keys: list[str] = list(set(tensor1.keys()) & set(tensor2.keys()))
    common_keys.remove("predictions")
    for key in common_keys:
        tensor_comparison[f"tensor1_{key}"] = tensor1[key]
        tensor_comparison[f"tensor2_{key}"] = tensor2[key]

    return tensor_comparison


def compare_tensors_and_save(
    tensor_paths: list[str],
    output_logs: str,
    mode: str = "cosine_similarity",
) -> None:
    """Compare tensors and save the results to a CSV file.

    Args:
        tensor_paths: List of paths to the tensors to compare.
        output_logs: Path to save the logs.
        mode: Mode to use for comparison.
    """
    results: dict[str, list[float | str]] = defaultdict(list)
    tensors: dict[str, dict[str, torch.Tensor]] = {path: load_file(path) for path in tensor_paths}
    for i in range(len(tensor_paths)):
        for j in range(i + 1, len(tensor_paths)):
            tensor1 = tensors[tensor_paths[i]]
            tensor2 = tensors[tensor_paths[j]]
            tensor_comparison = compare_tensors(tensor1, tensor2, mode)
            for key, tensor in tensor_comparison.items():
                if tensor.ndim == 0:
                    results[key].append(tensor.item())
                else:
                    results[key].append(tensor.mean().item())
    pl.DataFrame(results).write_csv(output_logs)
