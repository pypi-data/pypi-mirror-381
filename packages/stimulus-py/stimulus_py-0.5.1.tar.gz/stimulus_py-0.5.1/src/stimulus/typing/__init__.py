"""Typing for Stimulus Python API.

This module contains all Stimulus types which will be used for variable typing
and likely not instantiated, as well as aliases for other types to use for typing purposes.

The aliases from this module should be used for typing purposes only.
"""
# ruff: noqa: F401
# ruff: noqa: F811

from typing import Any, TypeAlias, TypeVar

# these imports mostly alias everything
# Note: DatasetHandler, DatasetLoader, DatasetProcessor, TorchDataset removed as they were unused
from stimulus.data.encoding.encoders import AbstractEncoder as Encoder
from stimulus.data.interface.data_config_parser import (
    create_encoders,
    create_splitter,
    create_transforms,
)
from stimulus.data.interface.data_config_schema import (
    Columns,
    ColumnsEncoder,
    ConfigDict,
    GlobalParams,
    Schema,
    Split,
    SplitConfigDict,
    SplitTransformDict,
    Transform,
    TransformColumns,
    TransformColumnsTransformation,
)
from stimulus.data.splitting import AbstractSplitter as Splitter
from stimulus.data.transforming.transforms import AbstractTransform as Transform
from stimulus.typing.protocols import StimulusModel

# Note: PredictWrapper and Performance imports removed due to missing modules
# Note: yaml_model_schema imports removed as they contained unused Ray-related classes
# The codebase now uses Optuna exclusively for hyperparameter tuning

# data/interface/data_config_schema.py

YamlData: TypeAlias = (
    Columns
    | ColumnsEncoder
    | ConfigDict
    | GlobalParams
    | Schema
    | Split
    | SplitConfigDict
    | Transform
    | TransformColumns
    | TransformColumnsTransformation
    | SplitTransformDict
)

# Replace these problematic imports
# Note: raytune_learner imports removed as the module was replaced with optuna_tune

# Replace with type aliases if needed
CheckpointDict = dict[str, Any]
TuneModel = TypeVar("TuneModel")
TuneWrapper = TypeVar("TuneWrapper")
