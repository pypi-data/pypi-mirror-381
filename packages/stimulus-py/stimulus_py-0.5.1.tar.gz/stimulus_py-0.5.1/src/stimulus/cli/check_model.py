#!/usr/bin/env python3
"""CLI module for checking model configuration and running initial tests."""

import logging
import os
from typing import Optional

import datasets
import optuna
import yaml

from stimulus.learner import optuna_tune
from stimulus.learner.device_utils import resolve_device
from stimulus.learner.interface import model_schema
from stimulus.utils import model_file_interface

logger = logging.getLogger(__name__)

MAX_SAMPLES = 1000
COMPUTE_OBJECTIVE_EVERY_N_SAMPLES = 100
N_TRIALS = 5


def check_model(
    data_path: str,
    model_path: str,
    model_config_path: str,
    optuna_results_dirpath: str = "./optuna_results",
    force_device: Optional[str] = None,
) -> tuple[str, str]:
    """Run the main model checking pipeline.

    Args:
        data_path: Path to input data file.
        model_path: Path to model file.
        model_config_path: Path to model config file.
        optuna_results_dirpath: Directory for optuna results.
        force_device: Force the device to use.
    """
    dataset_dict = datasets.load_from_disk(data_path)
    dataset_dict.set_format("torch")
    train_dataset = dataset_dict["train"]
    validation_dataset = dataset_dict["test"]
    logger.info("Dataset loaded successfully.")

    model_class = model_file_interface.import_class_from_file(model_path)

    logger.info("Model class loaded successfully.")

    with open(model_config_path) as file:
        model_config_content = yaml.safe_load(file)
        model_config = model_schema.Model(**model_config_content)

    logger.info("Model config loaded successfully.")

    base_path = optuna_results_dirpath
    artifact_path = optuna_results_dirpath + "/artifacts"
    os.makedirs(base_path, exist_ok=True)
    os.makedirs(artifact_path, exist_ok=True)
    artifact_store = optuna.artifacts.FileSystemArtifactStore(base_path=artifact_path)
    storage = optuna.storages.JournalStorage(
        optuna.storages.journal.JournalFileBackend(f"{base_path}/optuna_journal_storage.log"),
    )

    device = resolve_device(force_device=force_device, config_device=model_config.device)
    objective = optuna_tune.Objective(
        model_class=model_class,
        network_params=model_config.network_params,
        optimizer_params=model_config.optimizer_params,
        data_params=model_config.data_params,
        loss_params=model_config.loss_params,
        train_torch_dataset=train_dataset,
        val_torch_dataset=validation_dataset,
        artifact_store=artifact_store,
        max_samples=model_config.max_samples,
        compute_objective_every_n_samples=model_config.compute_objective_every_n_samples,
        target_metric=model_config.objective.metric,
        device=device,
    )

    logger.info(f"Objective: {objective}")
    study = optuna_tune.tune_loop(
        objective=objective,
        pruner=optuna.pruners.MedianPruner(n_warmup_steps=2, n_startup_trials=2),
        sampler=optuna.samplers.TPESampler(),
        n_trials=N_TRIALS,
        direction=model_config.objective.direction,
        storage=storage,
    )
    if study is None:
        raise ValueError("Study is None")
    logger.info(f"Study: {study}")
    logger.info(f"Study best trial: {study.best_trial}")
    logger.info(f"Study direction: {study.direction}")
    logger.info(f"Study best value: {study.best_value}")
    logger.info(f"Study best params: {study.best_params}")
    logger.info(f"Study trials count: {len(study.trials)}")

    for artifact_meta in optuna.artifacts.get_all_artifact_meta(study_or_trial=study):
        logger.info(artifact_meta)
    # Download the best model
    trial = study.best_trial
    best_artifact_id = trial.user_attrs["model_id"]
    file_path = trial.user_attrs["model_path"]
    optuna.artifacts.download_artifact(
        artifact_store=artifact_store,
        file_path=file_path,
        artifact_id=best_artifact_id,
    )

    logger.info(f"Best model downloaded successfully to {file_path}")

    return base_path, file_path
