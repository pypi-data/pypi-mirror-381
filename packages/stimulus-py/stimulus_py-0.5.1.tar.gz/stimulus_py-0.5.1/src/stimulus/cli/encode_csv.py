"""CLI module for encoding CSV data files."""

import logging
from typing import Any, Optional

import datasets
import numpy as np
import yaml

from stimulus.data.interface import data_config_parser
from stimulus.data.interface.data_loading import load_dataset_from_path

logger = logging.getLogger(__name__)


def load_encoders_from_config(data_config_path: str) -> dict[str, Any]:
    """Load the encoders from the data config.

    Args:
        data_config_path: Path to the data config file.

    Returns:
        A dictionary mapping column names to encoder instances.
    """
    with open(data_config_path) as file:
        data_config_dict = yaml.safe_load(file)
        data_config_obj = data_config_parser.EncodingConfigDict(**data_config_dict)

    encoders, _input_columns, _label_columns, _meta_columns = data_config_parser.parse_encoding_config(
        data_config_obj,
    )

    # Return all encoders for all column types
    return encoders


def encode_batch(
    batch: datasets.formatting.formatting.LazyBatch,
    encoders_config: dict[str, Any],
) -> dict[str, list]:
    """Encode a batch of data.

    This function applies configured encoders to specified columns within a batch.
    Each encoder's `batch_encode` method is called to transform the column data.

    Args:
        batch: The input batch of data (a Hugging Face LazyBatch).
        encoders_config: A dictionary where keys are column names and values are
                        encoder objects to be applied to that column.

    Returns:
        A dictionary representing the encoded batch, with all original columns
        present and encoded columns updated according to the encoders.
    """
    result_dict = dict(batch)

    for column_name, encoder in encoders_config.items():
        if column_name not in batch:
            logger.warning(
                f"Column '{column_name}' specified in encoders_config was not found "
                f"in the batch columns (columns: {list(batch.keys())}). Skipping encoding for this column.",
            )
            continue

        # Get the column data as numpy array
        column_data = np.array(batch[column_name])

        # Apply the encoder
        try:
            encoded_data = encoder.batch_encode(column_data)
            result_dict[column_name] = encoded_data.tolist() if isinstance(encoded_data, np.ndarray) else encoded_data
        except Exception:
            logger.exception(f"Failed to encode column '{column_name}'")
            raise

    return result_dict


def main(data_path: str, config_yaml: str, out_path: str, num_proc: Optional[int] = None) -> None:
    """Encode the data according to the configuration.

    Args:
        data_path: Path to input data (CSV, parquet, or HuggingFace dataset directory).
        config_yaml: Path to config YAML file.
        out_path: Path to output encoded dataset directory.
        num_proc: Number of processes to use for encoding.
    """
    # Load the dataset
    dataset = load_dataset_from_path(data_path)

    # Set format to numpy for processing
    dataset.set_format(type="numpy")

    # Load encoders from config
    encoders = load_encoders_from_config(config_yaml)
    logger.info("Encoders initialized successfully.")
    logger.info(f"Loaded encoders for columns: {list(encoders.keys())}")

    # Identify and remove columns that aren't in the encoder configuration
    encoder_columns = set(encoders.keys())
    columns_to_remove = set()

    for split_name, split_dataset in dataset.items():
        dataset_columns = set(split_dataset.column_names)
        split_columns_to_remove = dataset_columns - encoder_columns
        columns_to_remove.update(split_columns_to_remove)
        logger.info(f"Split '{split_name}' columns to remove: {list(split_columns_to_remove)}")

    if columns_to_remove:
        logger.info(f"Removing columns not in encoder configuration: {list(columns_to_remove)}")
        dataset = dataset.remove_columns(list(columns_to_remove))

    # Apply the encoders to the data
    dataset = dataset.map(
        encode_batch,
        batched=True,
        fn_kwargs={"encoders_config": encoders},
        num_proc=num_proc,
    )

    logger.info(f"Dataset encoded successfully. Saving to: {out_path}")

    # Save the encoded dataset to disk
    dataset.save_to_disk(out_path)

    logger.info(f"Encoded dataset saved to: {out_path}")
