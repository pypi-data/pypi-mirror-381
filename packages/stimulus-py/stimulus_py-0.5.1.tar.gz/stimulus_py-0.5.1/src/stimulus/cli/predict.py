#!/usr/bin/env python3
"""CLI module for model prediction on datasets."""

import json
import logging

import datasets
import safetensors.torch as safetensors
import torch

from stimulus.data.interface.data_loading import load_dataset_from_path
from stimulus.typing.protocols import StimulusModel
from stimulus.utils.model_file_interface import import_class_from_file

logger = logging.getLogger(__name__)


def load_model(model_path: str, model_config_path: str, weight_path: str) -> StimulusModel:
    """Dynamically loads the model from a .py file."""
    with open(model_config_path) as f:
        complete_config = json.load(f)

    # Extract network parameters from complete config
    # Handle both old format (direct params) and new format (nested params)
    network_params = complete_config.get("network_params", complete_config)

    # Check that the model can be loaded
    model = import_class_from_file(model_path)
    model_instance = model(**network_params)

    weights = safetensors.load_file(weight_path)
    model_instance.load_state_dict(weights)
    return model_instance


def update_statistics(statistics: dict, temp_statistics: dict) -> dict:
    """Update the statistics with the new statistics.

    Args:
        statistics: The statistics to update.
        temp_statistics: The new statistics to update with.

    Returns:
        The updated statistics.
    """
    for key, value in temp_statistics.items():
        if isinstance(value, torch.Tensor):
            if value.ndim == 0:
                try:
                    # Check if statistics[key] is zero-dimensional and reshape it if needed
                    if statistics[key].ndim == 0:
                        statistics[key] = torch.cat([statistics[key].reshape(1), value.unsqueeze(0)], dim=0)
                    else:
                        statistics[key] = torch.cat([statistics[key], value.unsqueeze(0)], dim=0)
                except RuntimeError as e:
                    raise RuntimeError(
                        f"Error updating statistics: {e}, shape of incoming tensors is {value.shape} and in-place tensor is {statistics[key].shape}, values of those tensors are {value} and {statistics[key]}",
                    ) from e
            else:
                statistics[key] = torch.cat([statistics[key], value], dim=0)
        elif isinstance(value, (int, float, list)):
            statistics[key] = statistics[key] + value
        else:
            raise TypeError(f"Invalid statistics type: {type(value)}")

    return statistics


def convert_dict_to_tensor(data: dict) -> dict:
    """Convert a dictionary to a tensor.

    Args:
        data: The dictionary to convert.

    Returns:
        The converted dictionary.
    """
    for key, value in data.items():
        if not isinstance(value, torch.Tensor):
            data[key] = torch.tensor(value)
    return data


def predict(
    data_path: str,
    model_path: str,
    model_config_path: str,
    weight_path: str,
    output: str,
    batch_size: int = 256,
) -> None:
    """Run model prediction pipeline.

    Args:
        data_path: Path to the input data file.
        model_path: Path to the model file.
        model_config_path: Path to the model config YAML file.
        weight_path: Path to the model weights file.
        output: Path to save the prediction results.
        batch_size: Batch size for prediction.
    """
    # Load model configuration to get loss function
    with open(model_config_path) as file:
        complete_config = json.load(file)

    # Get loss function from the optimized config
    loss_params = complete_config.get("loss_params", {})
    if not loss_params:
        raise ValueError(f"No loss_params found in model config: {model_config_path}")

    # Extract loss function name from optimized parameters
    # The loss function name should be directly available (e.g., {"loss_fn": "BCEWithLogitsLoss"})
    loss_fn_name = None
    for _param_name, param_value in loss_params.items():
        if isinstance(param_value, str):
            loss_fn_name = param_value
            break

    if not loss_fn_name:
        raise ValueError(f"Could not extract loss function from loss_params: {loss_params}")

    # Create loss function instance
    try:
        loss_fn = getattr(torch.nn, loss_fn_name)()
    except AttributeError as e:
        raise ValueError(f"Invalid loss function '{loss_fn_name}' in config") from e

    logger.info(f"Using loss function: {loss_fn_name}")

    # Get the best model with best architecture and weights
    model = load_model(model_path, model_config_path, weight_path)
    dataset = load_dataset_from_path(data_path)
    dataset.set_format(type="torch")
    splits = [dataset[split_name] for split_name in dataset]
    all_splits = datasets.concatenate_datasets(splits)
    loader = torch.utils.data.DataLoader(all_splits, batch_size=batch_size, shuffle=False)

    # create empty tensor for predictions
    is_first_batch = True
    for batch in loader:
        if is_first_batch:
            _loss, statistics = model.inference(batch, loss_fn)
            is_first_batch = False
        _loss, temp_statistics = model.inference(batch, loss_fn)
        statistics = update_statistics(statistics, temp_statistics)

    to_return: dict = convert_dict_to_tensor(statistics)
    # Predict the data
    safetensors.save_file(to_return, output)
