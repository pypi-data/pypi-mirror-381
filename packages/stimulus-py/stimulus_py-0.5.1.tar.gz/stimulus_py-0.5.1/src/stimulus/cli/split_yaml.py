#!/usr/bin/env python3
"""CLI module for splitting YAML configuration files into component configs.

This module provides functionality to split a single YAML configuration file into
separate component files: encoding config, individual split configs, and individual
transform configs. The resulting YAML files can be used independently.
"""

import logging
from pathlib import Path
from typing import Any

import yaml

from stimulus.data.interface.data_config_parser import ConfigDict, split_config_into_components

logger = logging.getLogger(__name__)


def split_yaml(config_yaml: str, out_dir_path: str) -> None:
    """Split a YAML config file into separate component configs.

    Takes a master YAML configuration and splits it into:
    - encode.yaml: Contains encoding configuration (global_params + columns)
    - split1.yaml, split2.yaml, etc.: Individual split configurations
    - transform1.yaml, transform2.yaml, etc.: Individual transform configurations
      with parameter expansion

    Args:
        config_yaml: Path to the master YAML configuration file.
        out_dir_path: Output directory to save the component config files.

    Raises:
        FileNotFoundError: If the config file doesn't exist.
        ValueError: If the YAML config is invalid or malformed.
    """
    # Validate input file exists
    config_path = Path(config_yaml)
    if not config_path.exists():
        raise FileNotFoundError(f"Configuration file not found: {config_yaml}")

    # Create output directory if it doesn't exist
    output_dir = Path(out_dir_path)
    output_dir.mkdir(parents=True, exist_ok=True)

    # Load and validate the YAML config
    yaml_config: dict[str, Any] = {}
    try:
        with open(config_yaml) as conf_file:
            yaml_config = yaml.safe_load(conf_file)
    except yaml.YAMLError as e:
        raise ValueError(f"Invalid YAML format in {config_yaml}: {e}") from e

    # Validate config structure
    try:
        config_dict = ConfigDict(**yaml_config)
    except Exception as e:
        raise ValueError(f"Invalid config structure in {config_yaml}: {e}") from e

    logger.info("YAML config loaded and validated successfully.")

    # Extract base name from config file path
    base_name = config_path.stem

    # Split the config into components
    split_config_into_components(config_dict, str(output_dir), base_name)

    # Count generated files for logging
    encoding_files = 1  # encode.yaml
    split_files = len(config_dict.split)
    transform_files = (
        sum(
            len(
                [
                    p
                    for p in transform.columns[0].transformations[0].params.values()
                    if isinstance(p, list) and len(p) > 1
                ],
            )
            if transform.columns
            and transform.columns[0].transformations
            and transform.columns[0].transformations[0].params
            else 1
            for transform in config_dict.transforms
        )
        if config_dict.transforms
        else 0
    )

    total_files = encoding_files + split_files + transform_files

    logger.info(f"Successfully generated {total_files} component configs:")
    logger.info(f"  - 1 encoding config ({base_name}_encode.yaml)")
    logger.info(f"  - {split_files} split configs ({base_name}_*_split.yaml)")
    logger.info(f"  - {transform_files} transform configs ({base_name}_*_transform.yaml)")
    logger.info(f"All files saved to: {out_dir_path}")
