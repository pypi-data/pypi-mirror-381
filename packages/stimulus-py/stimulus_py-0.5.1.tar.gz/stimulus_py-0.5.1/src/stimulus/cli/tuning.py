#!/usr/bin/env python3
"""CLI module for running Optuna hyperparameter tuning experiments."""

import logging
import os
import shutil
from typing import Any, Optional

import datasets
import optuna
import yaml

from stimulus.learner import optuna_tune
from stimulus.learner.device_utils import resolve_device
from stimulus.learner.interface import model_config_parser, model_schema
from stimulus.utils import model_file_interface

logger = logging.getLogger(__name__)


def tune(
    data_path: str,
    model_path: str,
    model_config_path: str,
    optuna_results_dirpath: str = "./optuna_results",
    best_model_path: str = "best_model.safetensors",
    best_optimizer_path: str = "best_optimizer.pt",
    best_config_path: str = "best_config.json",
    force_device: Optional[str] = None,
) -> None:
    """Run model hyperparameter tuning.

    Args:
        data_path: Path to input data file.
        model_path: Path to model file.
        model_config_path: Path to model config file.
        optuna_results_dirpath: Directory for optuna results.
        best_model_path: Path to write the best model to.
        best_optimizer_path: Path to write the best optimizer to.
        force_device: Force the device to use.
    """
    # Load train and validation datasets
    dataset_dict = datasets.load_from_disk(data_path)
    dataset_dict.set_format("torch")
    train_dataset = dataset_dict["train"]
    validation_dataset = dataset_dict["test"]

    # Load model class
    model_class = model_file_interface.import_class_from_file(model_path)

    # Load model config
    with open(model_config_path) as file:
        model_config_dict: dict[str, Any] = yaml.safe_load(file)
    model_config: model_schema.Model = model_schema.Model(**model_config_dict)

    # get the pruner
    pruner = model_config_parser.get_pruner(model_config.pruner)

    # get the sampler
    sampler = model_config_parser.get_sampler(model_config.sampler)

    # storage setups
    base_path = optuna_results_dirpath
    artifact_path = optuna_results_dirpath + "/artifacts"
    os.makedirs(base_path, exist_ok=True)
    os.makedirs(artifact_path, exist_ok=True)
    artifact_store = optuna.artifacts.FileSystemArtifactStore(base_path=artifact_path)
    storage = optuna.storages.JournalStorage(
        optuna.storages.journal.JournalFileBackend(f"{base_path}/optuna_journal_storage.log"),
    )

    device = resolve_device(force_device=force_device, config_device=model_config.device)

    objective = optuna_tune.Objective(
        model_class=model_class,
        network_params=model_config.network_params,
        optimizer_params=model_config.optimizer_params,
        data_params=model_config.data_params,
        loss_params=model_config.loss_params,
        train_torch_dataset=train_dataset,
        val_torch_dataset=validation_dataset,
        artifact_store=artifact_store,
        max_samples=model_config.max_samples,
        compute_objective_every_n_samples=model_config.compute_objective_every_n_samples,
        target_metric=model_config.objective.metric,
        device=device,
    )

    study = optuna_tune.tune_loop(
        objective=objective,
        storage=storage,
        sampler=sampler,
        pruner=pruner,
        n_trials=model_config.n_trials,
        direction=model_config.objective.direction,
    )

    best_trial = study.best_trial
    best_model_artifact_id = best_trial.user_attrs["model_id"]
    best_optimizer_artifact_id = best_trial.user_attrs["optimizer_id"]
    best_model_file_path = best_trial.user_attrs["model_path"]
    best_optimizer_file_path = best_trial.user_attrs["optimizer_path"]
    best_model_suggestions_artifact_id = best_trial.user_attrs["model_suggestions_id"]
    best_model_suggestions_file_path = best_trial.user_attrs["model_suggestions_path"]

    optuna.artifacts.download_artifact(
        artifact_store=artifact_store,
        file_path=best_model_file_path,
        artifact_id=best_model_artifact_id,
    )
    optuna.artifacts.download_artifact(
        artifact_store=artifact_store,
        file_path=best_optimizer_file_path,
        artifact_id=best_optimizer_artifact_id,
    )
    optuna.artifacts.download_artifact(
        artifact_store=artifact_store,
        file_path=best_model_suggestions_file_path,
        artifact_id=best_model_suggestions_artifact_id,
    )
    try:
        shutil.move(best_model_file_path, best_model_path)
        shutil.move(best_optimizer_file_path, best_optimizer_path)
        shutil.move(best_model_suggestions_file_path, best_config_path)
    except FileNotFoundError:
        logger.info("Best model or optimizer file_path not found, creating output directories")
        os.makedirs(os.path.dirname(best_model_path), exist_ok=True)
        os.makedirs(os.path.dirname(best_optimizer_path), exist_ok=True)
        os.makedirs(os.path.dirname(best_config_path), exist_ok=True)
        shutil.move(best_model_file_path, best_model_path)
        shutil.move(best_optimizer_file_path, best_optimizer_path)
        shutil.move(best_model_suggestions_file_path, best_config_path)
