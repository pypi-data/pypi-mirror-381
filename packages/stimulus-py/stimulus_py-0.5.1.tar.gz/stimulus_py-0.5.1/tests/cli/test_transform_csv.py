"""Tests for the transform_csv CLI command."""

import hashlib
import os
import pathlib
import tempfile
from typing import Any, Callable

import pytest
from click.testing import CliRunner

from stimulus.cli.main import cli
from stimulus.cli.transform_csv import main


# Fixtures
@pytest.fixture
def csv_path() -> str:
    """Fixture that returns the path to a CSV file."""
    return str(
        pathlib.Path(__file__).parent.parent / "test_data" / "titanic" / "titanic_stimulus.csv",
    )


@pytest.fixture
def yaml_path() -> str:
    """Fixture that returns the path to a YAML config file."""
    return str(
        pathlib.Path(__file__).parent.parent / "test_data" / "titanic" / "titanic_unique_transform.yaml",
    )


@pytest.mark.skip(reason="Non deterministic snapshot based on platform")
def test_transform_csv(
    csv_path: str,
    yaml_path: str,
    snapshot: Callable[[], Any],
) -> None:
    """Tests the transform_csv function with correct YAML files."""
    # Verify required files exist
    assert os.path.exists(csv_path), f"CSV file not found at {csv_path}"
    assert os.path.exists(yaml_path), f"YAML config not found at {yaml_path}"

    # Create temporary output file
    with tempfile.NamedTemporaryFile(suffix=".csv", delete=False) as tmp_file:
        output_path = tmp_file.name

    try:
        # Run main function - should complete without errors
        main(
            data_csv=csv_path,
            config_yaml=yaml_path,
            out_path=output_path,
        )

        # Verify output file exists and has content
        assert os.path.exists(output_path), "Output file was not created"
        with open(output_path, newline="", encoding="utf-8") as file:
            hash = hashlib.md5(file.read().encode()).hexdigest()  # noqa: S324
        assert hash == snapshot

    finally:
        # Clean up temporary file
        if os.path.exists(output_path):
            os.unlink(output_path)


@pytest.mark.skip(reason="Break github action runners")
def test_cli_invocation(
    csv_path: str,
    yaml_path: str,
) -> None:
    """Test the CLI invocation of transform-csv command."""
    runner = CliRunner()
    with runner.isolated_filesystem():
        output_path = "output.csv"
        result = runner.invoke(
            cli,
            [
                "transform-csv",
                "-c",
                csv_path,
                "-y",
                yaml_path,
                "-o",
                output_path,
            ],
        )
        assert result.exit_code == 0
        assert os.path.exists(output_path), "Output file was not created"
