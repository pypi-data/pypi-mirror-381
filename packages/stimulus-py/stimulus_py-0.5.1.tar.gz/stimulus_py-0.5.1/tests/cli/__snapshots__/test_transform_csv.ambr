# serializer version: 1
# name: test_transform_csv
  '5b558cb92b4187294f81bf6f1b059195'
# ---
